File: ReadyTasks/FSharp/ProjectEuler/Task_14/Task_14/Program.fs
// Learn more about F# at http://docs.microsoft.com/dotnet/fsharp

open System

let rec getCollatz numb = 
    if (numb = 1L) then
        1L
    else if (numb % 2L = 0L) then
        1L + (getCollatz (numb / 2L))
    else
        1L + (getCollatz ((3L * numb + 1L) / 2L))

let memorise func =
    let cache = new System.Collections.Generic.Dictionary<_,_>()
    fun x -> match cache.TryGetValue x with
    | true, result -> result
    | _ -> let result = func x
           cache.Add(x, result)
           result

let getCollatzMemorised = memorise getCollatz

let getResult numb =
    [1..numb] |> List.map int64 |> List.maxBy getCollatzMemorised

[<EntryPoint>]
let main argv =
    printfn "%d" (getResult 1000000)
    0 // return an integer exit code

File: ReadyTasks/CSharp/ProjectEuler/Task_14/Task_14/Program.cs
using System;
using System.Collections.Generic;

namespace Task_14
{
    class Program
    {
        static long GetResult(int max)
        {
            long result = 0;
            int maxSteps = 0;
            Dictionary<long, int> cache = new Dictionary<long, int>() { { 1, 0 } };
            for (long i = 2; i <= max; i++)
            {
                int steps = GetLongestCollatzChainLength(i, cache);
                cache.Add(i, steps);
                if (steps > maxSteps)
                {
                    maxSteps = steps;
                    result = i;
                }
            }
            return result;
        }
        static int GetLongestCollatzChainLength(long val, Dictionary<long, int> hashset = null)
        {
            int steps = 0;
            while (val > 1)
            {
                if (hashset != null && hashset.ContainsKey(val))
                {
                    return steps + hashset[val];
                }
                if (val % 2 == 0)
                {
                    val = val / 2;
                }
                else
                {
                    val = 3 * val + 1;
                }
                steps++;
            }
            return steps;
        }
        static void Main(string[] args)
        {
            Console.WriteLine(GetResult(1000000));
        }
    }
}
